### PR TITLE
ParsedFunction can accept another function

### LINKED ISSUE #12179: Allow the use of functions in `ParsedFunction`
## Rationale

When using a `ParsedFunction`, we need to be able to use another function as its input.

## Description

An example syntax would be:
```
[Functions]
  [./area_fn]
    type = PiecewiseLinear
    x = '0 1 2'
    y = '5 10 7
  [../]
  [./Dh_fn]
    type = ParsedFunction
    value = 'sqrt(4 * area / pi)'
    vars = 'area'
    vals = 'area_fn'
  [../]
[]
```

## Impact

New functionality.

### PR DESCRIPTION
**Pull request by andrsd**
Adding the possibility to couple a function into a ParsedFunction, similarly
like we do it for postprocessors and scalar variables

Closes #12179

### CODE REVIEW

**Comment by andrsd** [comment]
This is good for #MOOSEnewsletter

File: framework/src/functions/MooseParsedFunctionWrapper.C [hunk 1]
```diff
@@ -138,16 +139,32 @@ MooseParsedFunctionWrapper::initialize()
...
+      // Store the location of this variable
+      _fn_index.push_back(i);
+    }
+
     // Case when a Real is supplied, convert std::string to Real
     else
     {
       // Use istringstream to convert, if it fails produce an error, otherwise add the variable to
       // the _vals variable
       if (!(ss >> tmp) || !ss.eof())
```
Comment (permcody) [inline framework/src/functions/MooseParsedFunctionWrapper.C]: I know you didn't write this code, but we know now that this isn't sufficient to catch all edge cases. We have a utility that should be called here instead. Perhaps you can fix while you are in here?

File: framework/include/functions/MooseParsedFunctionWrapper.h [hunk 2]
```diff
@@ -104,6 +105,12 @@ class MooseParsedFunctionWrapper
   /// Vector of pointers to PP values
   std::vector<Real *> _scalar_vals;
 
+  /// Stores the relative location of variables (in _vars) that are connected to Functions
+  std::vector<unsigned int> _fn_index;
+
+  /// Vector of functions
+  std::vector<Function *> _fns;
```
Comment (permcody) [inline framework/include/functions/MooseParsedFunctionWrapper.h]: Not my favorite class scoped variable name 😦

**Review by permcody — CHANGES_REQUESTED** [review]

File: framework/include/functions/MooseParsedFunctionWrapper.h (hunk 2 above)
Comment (andrsd) [reply framework/include/functions/MooseParsedFunctionWrapper.h]: But, it nicely matches `_pp_vals` and `_pp_index` a few lines above 🙃

**Comment by andrsd** [comment]
I actually need to fix a few comments in the header file, because with the fix I made they are now lying.

**Comment by andrsd** [comment]
I did some refactoring in the `MooseParsedFunctionWrapper` - see 1bb7c82 for the actual description of things I did. I removed a lot of IMO useless comments from the code, because much better description is now in the header file.

**Review by permcody — APPROVED** [review]

**Pull request merged**